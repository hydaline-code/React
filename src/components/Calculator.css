
body {
  background-image: url('../img/img.jpg');
  background-size: cover;
  background-repeat: no-repeat;
  background-attachment: fixed;
}

h1 {
  text-align: center;
  color: rgb(88, 247, 247);
}

.calculator {
  width: 320px;
  margin: 1rem auto;
  border: 1px solid #ccc;
  border-radius: 5px;
  padding: 10px;
  text-align: center;
  background-color: #f4f4f4;
}

.display {
  font-size: 36px;
  margin-bottom: 10px;
  background-color: #ccc;
  border: 1px solid #0fabf4;
  text-align: right;
  border-radius: 5px;
  padding: 0.5rem;
  overflow: hidden;
}

button {
  width: 60px;
  height: 50px;
  font-size: 24px;
  padding: 10px;
  border: 1px solid #0fabf4;
  border-radius: 5px;
  cursor: pointer;
  background-color: #f4f0f0e7;
}

button.last {
  width: 120px;
  height: 50px;
}

.color {
  background: rgb(248, 162, 43);
}

.wide {
  width: 120px; /* Width of a single column */
}

button:hover {
  background-color: #e3e0e0;
}
