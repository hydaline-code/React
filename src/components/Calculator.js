import React, { useState } from 'react';
import { v4 as uuidv4 } from 'uuid';
import calculate from '../logic/calculate';
import './Calculator.css';

const Calculator = () => {
  const [displayValue, setDisplayValue] = useState('0');
  const [calculatorData, setCalculatorData] = useState({
    total: null,
    next: null,
    operation: null,
  });

  const handleButtonClick = (value) => {
    setCalculatorData((prevData) => {
      const newData = calculate(prevData, value);
      const { total, next, operation } = newData;
      const newDisplayValue = next || total || '0';
      setDisplayValue(newDisplayValue);
      return newData;
    });
  };
  const CalculatorButtons = ({ value, className }) => (
    <button type="button" onClick={() => handleButtonClick(value)} className={className}>
      {value}
    </button>
  );

  const buttonRows = [
    ['AC', '+/-', '%', 'รท'],
    ['7', '8', '9', 'x'],
    ['4', '5', '6', '-'],
    ['1', '2', '3', '+'],
    ['0', '.', '='],
  ];

  return (
    <div className="calculator">
      <div className="display">{displayValue}</div>
      <div className="buttons">
        {buttonRows.map((row) => (
          <div className="row" key={uuidv4()}>
            {row.map((button, columnIndex) => {
              let buttonClass = '';

              if (button === '0' && columnIndex === 0) {
                buttonClass = 'wide';
              } else if (columnIndex === row.length - 1) {
                buttonClass = 'color';
              }

              const buttonId = `button-${button}`;

              return (
                <CalculatorButtons
                  key={buttonId}
                  value={button}
                  className={buttonClass}
                />
              );
            })}
          </div>
        ))}
      </div>
    </div>
  );
};

export default Calculator;
