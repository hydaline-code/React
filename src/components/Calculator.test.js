import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Calculator from './Calculator';

describe('Calculator', () => {
  it('should render correctly', () => {
    const { container } = render(<Calculator />);
    expect(container).toMatchSnapshot();
  });

  it('should handle button click', () => {
    const { getByText } = render(<Calculator />);
    const button = getByText('7');
    fireEvent.click(button);
    expect(button.textContent).toBe('7');
  });

  it('should reset display to 0 when "AC" button is pressed after other buttons', () => {
    const { container } = render(<Calculator />);
    const display = container.querySelector('.display');
    // Simulate clicking a button other than "AC"
    const button = container.querySelector('.wide');
    fireEvent.click(button);
    // Simulate clicking the "AC" button
    const acButton = container.querySelector('.color');
    fireEvent.click(acButton);

    expect(display.textContent).toBe('0');
  });
});
