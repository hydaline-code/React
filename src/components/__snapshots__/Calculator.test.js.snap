// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator should render correctly 1`] = `
<div>
  <div
    class="calculator"
  >
    <h5>
      React Calculator App
    </h5>
    <div
      class="display"
    >
      0
    </div>
    <div
      class="buttons"
    >
      <div
        class="row"
      >
        <button
          class=""
          type="button"
        >
          AC
        </button>
        <button
          class=""
          type="button"
        >
          +/-
        </button>
        <button
          class=""
          type="button"
        >
          %
        </button>
        <button
          class="color"
          type="button"
        >
          รท
        </button>
      </div>
      <div
        class="row"
      >
        <button
          class=""
          type="button"
        >
          7
        </button>
        <button
          class=""
          type="button"
        >
          8
        </button>
        <button
          class=""
          type="button"
        >
          9
        </button>
        <button
          class="color"
          type="button"
        >
          x
        </button>
      </div>
      <div
        class="row"
      >
        <button
          class=""
          type="button"
        >
          4
        </button>
        <button
          class=""
          type="button"
        >
          5
        </button>
        <button
          class=""
          type="button"
        >
          6
        </button>
        <button
          class="color"
          type="button"
        >
          -
        </button>
      </div>
      <div
        class="row"
      >
        <button
          class=""
          type="button"
        >
          1
        </button>
        <button
          class=""
          type="button"
        >
          2
        </button>
        <button
          class=""
          type="button"
        >
          3
        </button>
        <button
          class="color"
          type="button"
        >
          +
        </button>
      </div>
      <div
        class="row"
      >
        <button
          class="wide"
          type="button"
        >
          0
        </button>
        <button
          class=""
          type="button"
        >
          .
        </button>
        <button
          class="color"
          type="button"
        >
          =
        </button>
      </div>
    </div>
  </div>
</div>
`;
